VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,989 @@
+https://repo1.maven.org/maven2/org/apache/tomee/apache-tomee/ 
+Note how the constraints are sorted:
+
+
+- "affects Apache TomEE 8.0.0-M1 - 8.0.1, Apache TomEE 7.1.0 - 7.1.2,
+  Apache TomEE 7.0.0-M1 - 7.0.7, Apache TomEE 1.0.0-beta1 - 1.7.5."
+
+- a normalized version range spec is:
+  ``vers:tomee/>=1.0.0-beta1|<=1.7.5|>=7.0.0-M1|<=7.0.7|>=7.1.0|<=7.1.2|>=8.0.0-M1|<=8.0.1``
+
+- alternatively, four ``vers`` express the same range, using one ``vers`` for
+  each vulnerable "branches": 
+  - ``vers:tomee/>=1.0.0-beta1|<=1.7.5``
+  - ``vers:tomee/>=7.0.0-M1|<=7.0.7``
+  - ``vers:tomee/>=7.1.0|<=7.1.2``
+  - ``vers:tomee/>=8.0.0-M1|<=8.0.1``
+
+Conversing Rubygems custom syntax for dependency on gem. Note how the
+pessimistic version constraint is expanded:
+
+- ``'library', '~> 2.2.0', '!= 2.2.1'``
+- the version range spec is: ``vers:gem/>=2.2.0|!= 2.2.1|<2.3.0``
+
+
+URI scheme
+~~~~~~~~~~
+
+The ``vers`` URI scheme is  an acronym for "VErsion Range Specifier".
+It has been selected because it is short, obviously about version and available
+for a future formal registration for this URI-scheme at the IANA registry.
+
+The URI scheme is followed by a colon ":".
+
+
+``<versioning-scheme>``
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``<versioning-scheme>`` (such as ``npm``, ``deb``, etc.) determines:
+
+- the specific notation and conventions used for a version string encoded in
+  this scheme. Versioning schemes often specify a version segments separator and
+  the meaning of each version segments, such as [major.minor.patch] in semver.
+
+- how two versions are compared as greater or lesser to determine if a version
+  is within or outside a range.
+
+- how a versioning scheme-specific range notation can be transformed in the
+  ``vers`` simplified notation defined here.
+
+By convention the versioning scheme **should** be the same as the ``Package URL``
+package type for a given package ecosystem. It is OK to have other schemes
+beyond the purl type. A scheme could be specific to a single package name.
+
+The ``<versioning-scheme>`` is followed by a slash "/".
+
+
+``<version-constraint>``
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+After the ``<versioning-scheme>`` and "/" there are one or more
+``<version-constraint>`` separated by a pipe "|". The pipe "|" has no special
+meaning beside being a separator.
+
+Each  ``<version-constraint>`` of this list is either a single ``<version>`` as
+in ``1.2.3`` for example or the combination of a ``<comparator>`` and a ``<version>`` as in
+``>=2.0.0`` using this syntax::
+
+    <comparator><version>
+
+A single version that means that a version equal to this version satisfies the
+range spec. Equality is based on the equality of two normalized version strings
+according to their versioning scheme. For most schemes, this is a simple string
+equality. But schemes can specify normalization and rules for equality such as
+``pypi`` with PEP440. 
+
+
+The special star "*" comparator matches any version. It must be used **alone**
```

Comment:
A constraint with a ``*`` comparator is not a _constraint_ at all. IMO a better way to indicate no constraints is to not have a constraint in the given version range.

Reply:
It's been useful to me to use `*` to mean "all versions of the package", and leaving out the constraint to mean either the abstract package (i.e. the "concept" of zlib, for lack of a better term), or in cases where the version isn't known.

For example, in [OSS Gadget](https://github.com/Microsoft/OSSGadget), we can reference packages like:
```
  pkg:cpan/Apache-ACEProxy      The latest version of Apache::ACEProxy (via cpan.org)
  pkg:cran/ACNE@0.8.0           Version 0.8.0 of ACNE (via cran.r-project.org)
  pkg:gem/rubytree@*            All versions of RubyTree (via rubygems.org)
```

Of course, the convention that "no version specifier means the latest" was one that we made up for our use case, but it's been convenient.